Conebeam cone fixes

adler-j opened the pull request:
Initial fixes for the conebeam fixes, needs some more I suppose. Now uses the generalized functions wherever possible.

moosmann commented:
@adler-j: What's the proper way to proceed? Checkout your branch and fix the remaining issues or merge your branch and then fix the issues in the base branch?

adler-j commented:
Please note that this is a pull request *into* the `issue-87__conebeam` branch, not master.

adler-j commented on examples/tomo/xray_trafo.py:
```diff
@@ -63,4 +63,6 @@
 # Shows a slice of the phantom, projections, and reconstruction
 discr_vol_data.show(indices=np.s_[:, :, 150])
 discr_proj_data.show(indices=np.s_[0, :, :])
+discr_proj_data.show(indices=np.s_[30, :, :])
```
Test code, remove

adler-j commented on examples/tomo/xray_trafo.py:
```diff
@@ -41,18 +41,18 @@
 dgrid = odl.uniform_sampling(dparams, [558, 558])
 geom = odl.tomo.CircularConeFlatGeometry(angle_intvl, dparams, src_rad,
-                                         det_rad, agrid, dgrid)
+                                         det_rad, agrid, dgrid, axis=[0, 0, 1])
```
Use default

adler-j commented on odl/solvers/util/partial.py:
```diff
@@ -138,10 +138,15 @@ def __init__(self, *args, **kwargs):
         self.args = args
         self.kwargs = kwargs
         self.fig = None
+        self.plot_every_nth = kwargs.pop('plot_every_nth', 1)
```
This shouldnt be in this branch but on master.

kohr-h replied:
And name is not so great, it implies something boolean.

adler-j replied:
Yes that bloody name, any ideas?

adler-j replied:
Made an issue #213

adler-j commented on odl/tomo/backends/astra_setup.py:
```diff
@@ -202,33 +202,40 @@ def astra_geom_to_vec(geometry):
+    angles = geometry.motion_grid
+    num_angles = geometry.motion_grid.ntotal
+
+    if isinstance(geometry, ConeBeamGeometry) and isinstance(geometry.detector, FlatDetector):
```
Note: This is the only way to properly check that we have a flat panel cone beam geometry.

kohr-h replied:
It's two aspects, so two checks. We may want to have something like `is_curved` in the detector class in the long run.

adler-j commented on odl/tomo/backends/astra_setup.py:
```diff
@@ -202,33 +202,40 @@ def astra_geom_to_vec(geometry):
-            # vectors[nn, 9] = 0
-            # vectors[nn, 10] = 0
-            vectors[ang_idx, 11] = det_pix_height
+        # Astra order, needed for data to match what we expect from astra.
```
How do we properly document this?

kohr-h replied:
I'd say point to their doc, but I have my doubts that this is how it will work. Any chance to make a quick description?

adler-j commented on odl/tomo/backends/astra_setup.py:
```diff
@@ -339,7 +347,7 @@ def astra_projection_geometry(geometry):
             det_col_count, angles)
 
-    elif isinstance(geometry, CircularConeFlatGeometry):
+    elif False and isinstance(geometry, CircularConeFlatGeometry):
```
Imo, remove this whole elif clause unless ASTRA optimizes for the `cone` geometry.

moosmann replied:
Agree. `cone_vec` will be used for helical and circular acquisition curves. Vector arguments will be used accordingly for the parallel 3D and fanbeam case. However, the parallel 2d geometry does not support vector arguments according to the documentation.

adler-j replied:
One important case is this in the astra doc for FDK (if we want to add an inverse):

[astra::FDK_CUDA](http://sourceforge.net/p/astra-toolbox/wiki/FDK_CUDA/)

> Supported geometries: only cone. (cone_vec is not supported.)

kohr-h replied:
If we *really* need FDK then we need a clause here. But that's another issue and not the topic here. Parallel beam is a different story again. I'd say delete it here.

moosmann commented on odl/tomo/geometry/conebeam.py:
```diff
@@ -111,24 +89,96 @@ def motion_grid(self):
+
+
+class AxisOrientedGeometry(object):
+
```
Should this class moved to `geometry` since it will be used also for the parallel 3D case?

adler-j replied:
Yes.

moosmann replied:
We should have a corresponding class for the 2D case which provides the rotation matrix.

kohr-h commented on odl/tomo/backends/astra_setup.py:
```diff
@@ -202,33 +202,40 @@ def astra_geom_to_vec(geometry):
+        newind = []
+        for i in range(4):
+            newind += [2 + 3*i, 1 + 3*i, 0 + 3*i]
+        vectors = vectors[:, newind]
```
Some words on this, maybe a link to the ASTRA doc on this if existent?

adler-j replied:
Not quite sure if it is, we also do some reordering a few rows up, this is empirical.

moosmann replied:
A link to the ASTRA doc won't really help since the documentation is only for MATLAB which uses Fortran ordering.

moosmann replied:
The PyASTRA doc largely uses the doc for MATLAB and is incorrect in several places (at least the last time I have checked it).

kohr-h replied:
Well, then we need own words, otherwise this code is just magic.

kohr-h commented on odl/tomo/backends/astra_setup.py:
```diff
@@ -352,7 +360,7 @@ def astra_projection_geometry(geometry):
             angles, source_origin, origin_det)
 
-    elif isinstance(geometry, HelicalConeFlatGeometry):
+    elif isinstance(geometry, ConeBeamGeometry):
```
What happened to the helical guy?

adler-j replied:
Subset of ConeBeamGeometry now.

kohr-h commented on odl/tomo/geometry/conebeam.py:
```diff
@@ -19,67 +19,54 @@
-
-    The motion parameter is the (1d) rotation angle parametrizing source and
-    detector positions.
+    A cone beam geometry is characterized by a source and a
```
an

kohr-h commented on odl/tomo/geometry/conebeam.py:
```diff
@@ -19,67 +19,54 @@
+        ndim : {1, 2}
+            number of dimensions of geometry
+        angle_intvl : 1d `IntervalProd`
+            Admissible angles.
```
No dot

kohr-h commented on odl/tomo/geometry/conebeam.py:
```diff
@@ -19,67 +19,54 @@
+        angle_intvl : 1d `IntervalProd`
+            Admissible angles.
+        detector : `Detector`
+            The detector to use.
```
No dot

kohr-h commented on odl/tomo/geometry/conebeam.py:
```diff
@@ -111,24 +89,96 @@ def motion_grid(self):
+        dpar : element of detector parameters `det_params`
+            Detector parameter at which to evaluate
+        normalized : `bool`, optional
+            If `True`, return a normalized (unit) vector. Default: `True`
```
Default clear from the signature, remove

adler-j commented on odl/tomo/geometry/conebeam.py:
```diff
@@ -19,67 +19,54 @@
+class ConeBeamGeometry(Geometry):
 
-    """Abstract 3d cone beam geometry.
+    """Abstract nd cone beam geometry.
```
We should use this class also for 1d fanbeam I suppose.

adler-j commented on odl/tomo/geometry/conebeam.py:
```diff
@@ -19,67 +19,54 @@
 
 
-class ConeBeamGeometry(with_metaclass(ABCMeta, Geometry)):
+class ConeBeamGeometry(Geometry):
```
This can be moved to the main geometry file since it is now abstract.

moosmann replied:
I see your point, but isn't it confusing not to have this class in the `conebeam` module?

kohr-h replied:
Depends on where we intend to put the fan beam geometry in the long run.

adler-j replied:
Thats kindoff why I wanted the rename to `DivergentBeamGeometry` since this does not have a strong 3d connotation.

moosmann replied:
Shall we merge the `conebeam` and `fanbeam` module and rename it to `divergent`? Then it is also similar to the `parallel` module.

kohr-h replied:
Sounds good to me.

kohr-h commented on odl/tomo/geometry/conebeam.py:
```diff
@@ -111,24 +89,96 @@ def motion_grid(self):
+            raise ValueError('detector parameter {} is not in the valid '
+                             'range {}.'.format(dpar, self.det_params))
+
+        vec = self.det_point_position(mpar, dpar) - self.src_position(mpar)
```
This vector is pointing in the opposite direction.

adler-j commented on odl/tomo/geometry/conebeam.py:
```diff
@@ -111,24 +89,96 @@ def motion_grid(self):
+    def __repr__(self):
+        """Return ``repr(self)``"""
+
+        # TODO
```
Fix

kohr-h commented on odl/tomo/geometry/conebeam.py:
```diff
@@ -111,24 +89,96 @@ def motion_grid(self):
+class AxisOrientedGeometry(object):
+
+    """ Mixin class for cone beam geometries oriented according to an axis.
+    """
```
Remove first whitespace and line break

moosmann commented on odl/tomo/geometry/conebeam.py:
```diff
@@ -429,11 +399,47 @@ def __init__(self, angle_intvl, dparams, src_radius, det_radius,
         axis : `int` or 3-element array, optional
             Defines the rotation axis via a 3-element vector or a single
             integer referring to a standard axis. Default: `None`
```
Remove single integer option

adler-j commented on odl/tomo/geometry/geometry.py:
```diff
@@ -125,21 +125,7 @@ def det_to_src(self, mpar, dpar, normalized=True):
-            The source position, a `ndim`-dimensional vector
-        """
-        raise NotImplementedError
+        return NotImplemented
```
raise NotImplementedError

adler-j commented on odl/tomo/geometry/parallel.py:
```diff
@@ -308,7 +308,7 @@ def __init__(self, angle_intvl, dparams, agrid=None, dgrid=None,
                     len(axis), axis))
         self._axis = np.array(axis) / np.linalg.norm(axis)
-        self._detector = Flat2dDetector(dparams, dgrid)
+        self._detector = Flat2dDetector(dparams, [[0, 1, 0], [0, 0, 1]], dgrid)
```
Make sure this isn't the only place we need this.

adler-j replied:
Perhaps do this after we merge.

kohr-h replied:
That may very well end up as the only place we use this internally. Nothing bad about this, I guess, if it offers flexibility to the user.

kohr-h commented on odl/tomo/geometry/conebeam.py:
```diff
@@ -111,24 +89,96 @@ def motion_grid(self):
+        Parameters
+        ----------
+        axis : 3-element array, optional
+            Defines the rotation axis via a 3-element vector.
```
Better "Fixed rotation axis defined by a 3-element vector" (without dot)
Write what it *is*, not what it *does*.

kohr-h commented on odl/tomo/geometry/conebeam.py:
```diff
@@ -111,24 +89,96 @@ def motion_grid(self):
+            Defines the rotation axis via a 3-element vector.
+        """
+
+        self._axis = np.array(axis) / np.linalg.norm(axis)
```
Mention in the docstring that it's normalized

adler-j replied:
It says so:

"The normalized axis of rotation." in the `axis` property.

kohr-h commented on odl/tomo/geometry/conebeam.py:
```diff
@@ -210,7 +247,9 @@ class ConeFlatGeometry(ConeBeamGeometry):
-                 agrid=None, dgrid=None, axis=None):
+                 agrid=None, dgrid=None,
+                 axis=[0, 0, 1], source_to_detector=[1, 0, 0],
+                 detector_axises=None):
```
axes

kohr-h replied:
Quite a lot of arguments, consider moving some into `**kwargs`

adler-j replied:
Agree, but I'll leave it as is now. We can have an look at these later.

kohr-h commented on odl/tomo/geometry/conebeam.py:
```diff
@@ -224,72 +263,51 @@ def __init__(self, angle_intvl, dparams, src_radius, det_radius,
-            integer referring to a standard axis
+            A sampling grid for ``dparams``. Default: `None`
+        axis : 3-element array, optional
+            Defines the rotation axis via a 3-element vector.
```
~~Defines~~

kohr-h commented on odl/tomo/geometry/conebeam.py:
```diff
@@ -224,72 +263,51 @@ def __init__(self, angle_intvl, dparams, src_radius, det_radius,
+        axis : 3-element array, optional
+            Defines the rotation axis via a 3-element vector.
+        source_to_detector : 3-element array, optional
+            Defines the direction from the source to the point (0,0) of the
```
~~Defines the~~

kohr-h replied:
Normalized?

adler-j replied:
Not needed, code does it

kohr-h commented on odl/tomo/geometry/conebeam.py:
```diff
@@ -224,72 +263,51 @@ def __init__(self, angle_intvl, dparams, src_radius, det_radius,
+        source_to_detector : 3-element array, optional
+            Defines the direction from the source to the point (0,0) of the
+            detector.
+        detector_axises : tuple of two 3-element arrays, optional
```
sequence

kohr-h commented on odl/tomo/geometry/conebeam.py:
```diff
@@ -224,72 +263,51 @@ def __init__(self, angle_intvl, dparams, src_radius, det_radius,
+            Defines the direction from the source to the point (0,0) of the
+            detector.
+        detector_axises : tuple of two 3-element arrays, optional
+            Defines the unit directions along each detector parameter of the
```
~~Defines the~~

kohr-h commented on odl/tomo/geometry/conebeam.py:
```diff
@@ -406,7 +374,9 @@ class HelicalConeFlatGeometry(ConeFlatGeometry):
-                 pitch, agrid=None, dgrid=None, axis=(1, 0, 0)):
+                 pitch, agrid=None, dgrid=None,
+                 axis=[0, 0, 1], source_to_detector=[1, 0, 0],
+                 detector_axises=None):
```
See above

kohr-h commented on odl/tomo/geometry/conebeam.py:
```diff
@@ -429,11 +399,47 @@ def __init__(self, angle_intvl, dparams, src_radius, det_radius,
+            detector.
+        detector_axises : tuple of two 3-element arrays, optional
+            Defines the unit directions along each detector parameter of the
+            detector.
```
see above

kohr-h commented on odl/tomo/geometry/conebeam.py:
```diff
@@ -308,11 +326,11 @@ def det_refpoint(self, angle):
 
-        # ASTRA 'cone_vec' convention
-        return self.det_radius * np.array([-np.sin(angle), np.cos(angle), 0])
+        offset = self.det_radius * self._source_to_detector
```
Name "offset" is slightly confusing. This is supposed to be the not-yet-rotated reference point, there must be some word for this, maybe `initial_refpoint`.

kohr-h commented on odl/tomo/geometry/conebeam.py:
```diff
@@ -467,11 +473,14 @@ def det_refpoint(self, angle):
-            self.det_radius * np.cos(angle),
-            self.pitch * angle / (2 * np.pi)])
+        # Distance from 0 to detector
+        offset = self.det_radius * self._source_to_detector
```
Maybe it's me, but "offset" confuses me. I read an offset as a constant, once-for-all shift, the zero position everything else starts from. Why not write something like `circle_component` and `pitch_component`? That would anticipate that you intend to add them later.

kohr-h commented on odl/tomo/geometry/detector.py:
```diff
@@ -51,18 +51,18 @@ class Detector(with_metaclass(ABCMeta, object)):
 
         Parameters
         ----------
+        ndim : `int` in (0, 1 or 2)
```
`{0, 1, 2}`

kohr-h replied:
Why any restriction? Who says that 3- or 4-dimensional detectors are impossible? I'd just say `nonnegative int`.

adler-j replied:
We are OK with zero dimensional I suppose.

moosmann commented on odl/tomo/geometry/conebeam.py:
```diff
@@ -111,24 +89,96 @@ def motion_grid(self):
+        pos : `numpy.ndarray`, shape (`ndim`,)
+            The source position, a `ndim`-dimensional vector
+        """
+        raise NotImplementedError
```
We don't need this line since it is an abstract method, or do we?

kohr-h commented on odl/tomo/geometry/detector.py:
```diff
@@ -276,23 +279,31 @@ class Flat2dDetector(FlatDetector):
         params : `Rectangle` or 2-dim. `IntervalProd`
             The range of the parameters defining the detector area.
+        detector_axises : List with two tuples of 3 floats
+            The directions of the axises of the detector
```
See above

kohr-h commented on odl/tomo/operators/xray_trafo.py:
```diff
@@ -79,9 +79,13 @@ def __init__(self, discr_dom, geometry, backend='astra', **kwargs):
+
+        if not geometry.has_det_sampling:
+            raise ValueError('geometry {} does not have sampling grids for '
+                             'detector.'.format(geometry))
```
a sampling grid for the detector

moosmann commented on odl/tomo/geometry/conebeam.py:
```diff
@@ -184,22 +234,9 @@ def rotation_matrix(self, angle):
-
-class ConeFlatGeometry(ConeBeamGeometry):
-    """Cone beam geometry in 3d with flat detector.
+class CircularConeFlatGeometry(ConeBeamGeometry, AxisOrientedGeometry):
```
This class is pretty much the same as the helical class except with `pitch=0`. Why not making this a subclass of the helical class?

adler-j replied:
Did consider it, I'll get it done.

kohr-h commented:
Rebase from master to get the newest version of the submarine phantom. And it would be nice to get rid of the "Merge ..." commit. I'd say do a `rebase -i` and remove that merge commit, then rebase from master.

moosmann commented:
@adler-j: I'm going to work on the comments, or did you already anything?

adler-j commented:
Im doing work on them all right now, feels like my job.

moosmann commented:
OK.

adler-j commented:
Added examples for all the geometry cases, currently the parallel beam does not agree with the cone beam, I'll look into it.

moosmann commented:
You probably have seen it already:  `astra_parallel_3d_geom_to_vec`  still uses sin and cos instead the methods provided by the geometry.

adler-j commented:
Yes it has a TODO

moosmann commented on odl/tomo/geometry/detector.py:
```diff
@@ -276,23 +278,35 @@ class Flat2dDetector(FlatDetector):
         ----------
         params : `Rectangle` or 2-dim. `IntervalProd`
             The range of the parameters defining the detector area.
+        detector_axes : List with two tuples of 3 floats
```
sequence of two 3-element arrays

moosmann commented on odl/tomo/geometry/detector.py:
```diff
@@ -276,23 +278,35 @@ class Flat2dDetector(FlatDetector):
         params : `Rectangle` or 2-dim. `IntervalProd`
             The range of the parameters defining the detector area.
+        detector_axes : List with two tuples of 3 floats
+            The directions of the axises of the detector
```
axes

adler-j commented:
Several improvements, fanbeam and conebeam are now special cases of divergentbeam, which is general. One test fails since astra does not support flatfan_vec on the CPU, we could add a check to see if we are compatible with flatfan and use it in this case I suppose.

adler-j commented:
Having some issues with ASTRA parallel3d_vec, think they have a bug. Will commit then last review and we can merge this.

adler-j commented:
Commited a "fix" for it, cant figure out whats wrong with astra so cant even get a proper warning in there, we'll have to do with this one until ASTRA solves it on their side. It seems to be quite a "big" bug with `parallel3d_vec` since I get it with most of my example I try.

See [the issue on astra](https://github.com/astra-toolbox/astra-toolbox/issues/18) for any developments.

moosmann commented:
It works when rotation about [1,0,0]. Shall we simply enforce rotations about this axis for parallel 3D unless ASTRA has fixed the bug?

adler-j commented:
> It works when rotation about [1,0,0]. Shall we simply enforce rotations about this axis for parallel 3D unless ASTRA has fixed the bug?

Dont think having two different conventions for an axis is the best way to go either. Since this is astras fault imo we should simply warn about it and wait for them to fix it.

adler-j commented:
WJP has answered the issue and a fix should be out soon, so I wont put more time into that.

moosmann commented on odl/tomo/backends/astra_setup.py:
```diff
@@ -314,22 +371,20 @@ def astra_projection_geometry(geometry):
+            'fanflat_vec', det_count, vec)
 
-    elif isinstance(geometry, Parallel3dGeometry):
+    elif False and isinstance(geometry, Parallel3dGeometry):
```
This can be removed.

adler-j replied:
:+1:

moosmann commented on odl/tomo/backends/astra_setup.py:
```diff
@@ -202,79 +299,39 @@ def astra_geom_to_vec(geometry):
+        newind += [2 + 3 * i, 1 + 3 * i, 0 + 3 * i]
+    vectors = vectors[:, newind]
+
+    if np.any(vectors[:, 0:3] == 0):
```
I just tested the with the latest ASTRA version and it works. So can be removed.

moosmann commented on odl/tomo/geometry/parallel.py:
```diff
@@ -286,97 +268,50 @@ def __init__(self, angle_intvl, dparams, agrid=None, dgrid=None,
+            angle=0
+        detector_axes : sequence of two 3-element arrays, optional
+            Unit directions along each detector parameter of the detector.
+            Default: (normalized) [np.cross(axis, source_to_detector), axis]
```
source_to_detector -> origin_to_detector

moosmann commented on odl/tomo/backends/astra_setup.py:
```diff
@@ -470,9 +523,9 @@ def astra_projector(vol_interp, astra_vol_geom, astra_proj_geom, ndim, impl):
-                         'cone_vec'):
-        raise ValueError('invalid 2d geometry type {!r}.'.format(proj_type))
+    if proj_type not in ('parallel', 'fanflat', 'fanflat_vec',
+                         'parallel3d', 'parallel3d_vec', 'cone', 'cone_vec'):
```
Remove 'parallel3d', 'cone', and 'fanflat'.

adler-j replied:
Dont see why, code still works. If we want to remove it we should remove more code.

moosmann replied:
Right. We don't use these ASTRA geometries anymore, but here we don't need to change it.

moosmann commented on odl/tomo/geometry/conebeam.py:
```diff
@@ -28,97 +27,84 @@
+    and a pitch``P``. The detector reference point is opposite to
+    the source and moves on a spiral with radius ``R`` in the azimuthal
+    plane and pitch ``P``. The detector is aligned tangential to the
+    circle.
```
Remove last line.

moosmann commented on odl/tomo/geometry/conebeam.py:
```diff
@@ -28,97 +27,84 @@
-        return self._motion_params
+        if detector_axes is None:
+            detector_axes = [np.cross(self.axis, self._src_to_det),
+                             self.axis]
```
Ad check if `self.axis` and `self._src_to_det` are parallel?

adler-j replied:
Added a test for this inside the detector instead.

adler-j commented:
Fixed comments, merge?

moosmann commented:
Yes. Let's merge it.

adler-j commented:
Added default values for the "source_to_detector" thingie by adding a `perpendicular_vector` method. Also fixed some tests, merging after travis is happy.

adler-j commented:
Merged!
